sqlglot/optimizer/simplify.py

```diff
@@ -369,6 +372,50 @@ def absorb_and_eliminate(expression, root=True):
     return expression
 
 
+def propagate_constants(expression, root=True):
+    """
+    Propagate constants for conjunctions in DNF:
+
+    SELECT * FROM t WHERE a = b AND b = 5 becomes
+    SELECT * FROM t WHERE a = 5 AND b = 5
+
+    Reference: https://www.sqlite.org/optoverview.html
+    """
+
+    if (
+        isinstance(expression, exp.And)
+        and (root or not expression.same_parent)
+        and sqlglot.optimizer.normalize.normalized(expression, dnf=True)
+    ):
+        constant_mapping = {}
+        for eq in find_all_in_scope(expression, exp.EQ):
+            l, r = eq.left, eq.right
+
+            # TODO: create a helper that can be used to detect nested literal expressions such
+            # as CAST(123456 AS BIGINT), since we usually want to treat those as literals too
+            if isinstance(l, exp.Column) and isinstance(r, exp.Literal):
+                pass
+            elif isinstance(r, exp.Column) and isinstance(l, exp.Literal):
+                l, r = r, l
+            else:
+                continue
+
+            constant_mapping[l] = (id(l), r)
```

Comment:
why not put id(l) in the hash? that's most expensive part

Reply:
Not fully sure what you mean, can you elaborate? The intent was to use the hash to find all the matching columns (i.e. different `Expression` objects) in the conjunction by doing a lookup, and `id(l)` needs to be part of the stored value because we need to avoid replacing the actual column participating in the equality with the constant in that equality (i.e. don't replace `x` with `10` in `x = 10`).